R/collapse_counts.R

```diff
@@ -1,23 +1,71 @@
+#' 
+#' This function checks that all the expected flowcells are present in a table of detected flowcells.
+#' There can be more detected flowcells than there are expected flowcells.
+#' 
+#' @param detected_flowcells A dataframe with the columns "flowcell_name" and "flowcell_lane".
+#' @param expected_flowcells A dataframe with the columns "flowcell_name" and "flowcell_lane".
+validate_num_bio_reps= function(max_bio_rep_id, max_bio_rep_count) {
+  if(max_bio_rep_id > 1 & max_bio_rep_count == 1) {
+    stop('Unable to collapse bio_reps over the specified sig_cols.')
+  } else if(max_bio_rep_id < max_bio_rep_count) {
+    stop('Bio_reps were incorrectly collapsed resulting in more replicates than specified.')
+  } else if(max_bio_rep_id > 1 & max_bio_rep_id > max_bio_rep_count) {
+    print('Warning - Number of replicates that were collapses is smaller than the expected number of replicates.')
+    print('This could be due to a problem in processing or from poor data/sequencing quality.')
+  } else {}
+}
+
 #' collapse_counts
 #' 
 #' collapses l2fc values and computes MAD/sqrt(n) metrics for treatment conditions
 #'
 #'  @param l2fc - l2fc table with MAD/sqrt(n) metric for control condition
 #'  @return - collapsed_counts 
 #'  @export 
-collapse_counts = function(l2fc) {
+collapse_counts = function(l2fc, sig_cols) {
+  # Potentially static columns?
+  static_cols= c('project_code', 'CCLE_name', 'DepMap_ID', 'sig_id')
+  
+  # Validation: Check that sig_cols are present in l2fc.
+  if(validate_columns_exist== FALSE) {
+    print(sig_cols)
+    stop('Not all sig_cols (printed above) are present in the l2fc file.')
+  }
+  
+  # Median collapsing bio replicates.
   collapsed_counts = l2fc %>% 
     dplyr::filter(is.na(counts_flag)) %>% 
-    dplyr::group_by_at(setdiff(names(.), c('bio_rep', 'mean_n','mean_normalized_n', 'num_tech_reps', 'control_median_n',
-                                           'control_median_normalized_n', 'control_mad_sqrtN', 'num_ctrl_bio_reps', 
-                                           'control_MAD_QC','l2fc', 'counts_flag'))) %>% 
+    dplyr::group_by(pick(any_of(c(static_cols, sig_cols))) %>%
     dplyr::summarise(trt_median_n= median(mean_n), trt_median_normalized_n= median(mean_normalized_n),
                      trt_mad_sqrtN= mad(log2(mean_normalized_n)) / sqrt(dplyr::n()),
-                     median_l2fc= median(l2fc), num_bio_reps= dplyr::n()) %>% 
-    dplyr::ungroup() %>% 
+                     median_l2fc= median(l2fc), num_bio_reps= dplyr::n()) %>% dplyr::ungroup() %>% 
     dplyr::mutate(trt_MAD_QC= ifelse(trt_mad_sqrtN > 0.5/log10(2), F, T)) %>% # New: adjusted cut off to log2
     dplyr::relocate(trt_median_n, trt_median_normalized_n, trt_mad_sqrtN, 
                     num_bio_reps, median_l2fc, trt_MAD_QC, .after=last_col())
-  
+    
+    # Validation: Check that replicates were collapsed.
+    if('bio_rep' %in% colnames(l2fc)) {
+      max_bio_rep_id= max(unique(l2fc$bio_rep))
```

Comment:
Out of scope of this pull request, but just something to add to the metadata validation module in the future.

Specifically, the bio_rep column is numeric and can not contain NAs for this validation to make sense.